pkg/storage/backend-storage/backend-storage.go

```diff
@@ -45,8 +45,11 @@ import (
 )
 
 const (
-	PVCPrefix = "persistent-state-for"
-	PVCSize   = "10Mi"
+	// BackendPVCSizeAnnotation allows specifying a specific size for the backend PVC
+	BackendPVCSizeAnnotation = "storage.kubevirt.io/backend-pvc-size"
```

Comment:
```suggestion
	BackendPVCSizeAnnotation = "minimum-pvc-size"
```
Isn't it what it comes down to? I'd argue backend-storage or even KubeVirt doesn't have anything to do with the CSI (wrongly) requiring a minimum PVC size.

Reply:
@jean-edouard thanks for the review! The idea is to let users directly set the backend PVC size without implying it's just a minimum. This behavior is even more agnostic to the CSI requiring minimum size IMO since we are just letting users define the size. If we're not doing anything dynamic to adjust or enforce limits, calling it "minimum" in the annotation feels off.

As for `storage.kubevirt.io`, I'm unaware if we have a strict naming convention, but feels more consistent with other KubeVirt annotations.

wdyt?

Reply:
Yeah I agree, and the annotation is in the storage class. I get that ideally users shouldn't be concerned with the backend PVC size, but since we want to support this workaround, I consider `storage.kubevirt.io/backend-pvc-size` to be more descriptive and agnostic than `minimum-pvc-size`, since the annotation is not used to set a minimum, but rather the storage size itself.

Reply:
I agree with @jean-edouard's point here.

Firstly, it's important to mention that introducing a user-facing API via an annotation is more or less equivalent to adding an "official" API, since we'll have to keep backward compatibility when users would start using it. What are the advantages to using an annotation instead of a standard Kubevirt API?

Secondly, as @jean-edouard have said above, I'm not sure users care about the backend volume size, and I'm not even sure Kubevirt should care. The issue of not knowing the minimum PVC size for a storage class sounds like a general problem to me that should either have a general solution or that should not be solved at the Kubevirt level. Isn't there a standard / programmatic way to know the minimum PVC size?

/hold

Reply:
>  Isn't there a standard / programmatic way to know the minimum PVC size?

There isn’t afaik, this tends to be provisioner-specific, so solving it on our side would require implementing something similar to CDI's `StorageProfiles` to track minimum sizes per provisioner.

> What are the advantages to using an annotation instead of a standard Kubevirt API?

I agree that most users shouldn't need to care about the backend volume size so adding a spec field for this is overkill, but we are aware of users who can't use their VMs due to this limitation. This is the only storage size we are hardcoding so despite being far from the ideal solution, introducing an annotation as a workaround seems like a practical middle ground.

Also, whether users would care or not was already discussed offline with @jean-edouard and we seemed to agree that it’s necessary to provide a workaround to create backend storage PVCs while the size restriction is addressed per provisioner. Several users are waiting for this and I believe it’s more important to make the feature usable for them in the meantime. 
Pinging @dominikholler and @aglitke as they were part of the discussion.

Reply:
Here to voice my support for the annotation as a workaround.  Annotations are specifically not APIs and we have no obligation to continue supporting them.  Storage is a notoriously difficult area to provide general compatibility because even though there is the CSI standard, it leaves lots of room for differing behavior (as we see with AccessMode/VolumeMode differences and differing support characteristics for clone/snapshot).  We have been successful explaining to vendors that small PVC request sizes should not be rejected according to published kubernetes documentation but still we see several provisioners with this issue.  The question is if we want kubevirt to be usable with these provisioners until they can fix it on their side.

Reply:
> So might as well do it as a more general name such as: storage.kubevirt.io/minimal-size-request-supported

+1

Reply:
Thanks for the reviews! Following this conversation, I updated the annotation name to be `storage.kubevirt.io/minimum-supported-pvc-size`. Also added some logic so, if the used size is smaller than the default 10Mi, we use the 10Mi instead.

Reply:
> Annotations are specifically not APIs and we have no obligation to continue supporting them.

@aglitke I don't believe this statement is accurate.
What makes something a public API is that it is accessible for users to configure. If users would start relying on an annotation for their regular flows we would definitely need to take backward compatibility into considerations and won't be able to break it without a plan.

The bottom line from my POV is that we should discuss this like an API change, not as a workaround.

> making such a specific annotation on StorageClasses is strange since you are putting a very small kubevirt specific detail on a storage class which is widely used also for other stuff.

@ShellyKa13 I agree.
I wonder what's the advantage of using an annotation over having a config API at the kubevirt CR level. If we do that the admin can configure everything in a single place that won't affect public objects that are not part of Kubevirt.

Another thought about this is that I think a setup where one persona has RBAC permissions to create/edit storage classes and a different persona has permissions to create/edit Kubevirt makes sense.

> There isn’t afaik, this tends to be provisioner-specific, so solving it on our side would require implementing something similar to CDI's StorageProfiles to track minimum sizes per provisioner.

I just find it hard to believe that this is a kubevirt-specific problem. It sounds pretty generic.
I wonder how other projects are coping with this.

Perhaps a trial-and-error approach makes sense here? That is, we'll try to create a PVC with different sizes until we succeed, then write that we found in a ConfigMap or something similar?

------

I wonder if this demands a VEP that will expose this API slowly in order to do things right.

Reply:
FYI @vladikr

Reply:
Thanks for the input, @iholder101... yeah I agree with you that this is fundamentally a problem on the provisioner's side, and ideally they should handle this themselves. From that perspective, I think designing and maintaining a full-blown API on our side might be overkill for what's effectively a workaround to cope with non-compliant behavior.

If we don't want to go the annotation route, the only thing we might want to consider is something like what @arnongilboa proposed in the last sig-storage meeting, tracking minimum required sizes in a new field in the storage profiles. I think it wouldn't require much work since we are already using storage profiles in Kubevirt (even when creating backend storage PVCs) and is still discoverable/configurable in a centralized way.

That being said, I still like the annotation workaround, feels like a small and non-disruptive workaround meant to help some users (like those migrating from vSphere) get unblocked. If things change later, it’s much easier to drop or phase out an annotation than a spec field or a proper API.

Regarding the trial and error approach, seems to me way more complicated and hacky than using the annotation or storage profile to address specific provisioner issues.

Reply:
> tracking minimum required sizes in a new field in the storage profiles. I think it wouldn't require much work

> That being said, I still like the annotation workaround

You're talking about an automated process, right?
If that's not a lot of work and can be fully automated, what's the downside of this? Why do you prefer the annotation approach?

> If things change later, it’s much easier to drop or phase out an annotation than a spec field or a proper API.

I have deep problems with that statement :)
Essentially, there is no difference between a field and an annotation, as both are API, except for annotations being more hidden. If dropping an annotation's support or changing its behavior will break users, we need to consider backward compatibility just like we do with fields or any different form of API. The only easy way to remove an API is if it's not graduated (i.e. alpha/beta).

The fact that we're discussing removing the annotation before introducing it smells to me like we might come up with a better solution, especially if you say that one was proposed and shouldn't be a lot of work.

Reply:
> You're talking about an automated process, right?
If that's not a lot of work and can be fully automated, what's the downside of this? Why do you prefer the annotation approach?

Well, not a lot of work compared to designing and implementing an API from scratch but still would require maintaining and regularly updating the minimum volume sizes in the storage profile capabilities. As we pointed out during the meeting, owning and maintaining an API like this with hardcoded provisioner-specific values kind of ties us to a broken behavior. On other hand, the annotation feels more like a workaround that acknowledges the issue with the provisioner while still giving users a way to get their VMs working, especially in some corner cases where this becomes a blocker.

/cc @arnongilboa @akalenyu

Reply:
Hey @iholder101, just wanted to add that we have users blocked by this, and merging this solution would help unblock them, even if it's not ideal for the project. Would adding a beta or alpha prefix to the annotation help make it feel more appropriate?